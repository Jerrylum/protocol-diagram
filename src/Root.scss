@use "style/Essential.scss";
@use "style/Modals.scss";

#root {
  overflow: hidden;

  --bg-body-color: #ffffff;
  --bg-default-color: #ffffff;
  --bg-paper-color: #f3f3f3;
  --text-primary-color: rgba(0, 0, 0, 0.87);
  --text-disabled-color: rgba(0, 0, 0, 0.6);
  --link-color: rgba(0, 0, 0, 0.87);
  --hover-color: rgba(0, 0, 0, 0.08);
  --focused-color: rgba(0, 0, 0, 0.12);
  background-color: var(--bg-body-color);

  > #root-container {
    display: flex;
    width: 100vw;
    width: 100svw;
    height: 100vh;
    height: 100svh;
    box-sizing: border-box;
  }
}
