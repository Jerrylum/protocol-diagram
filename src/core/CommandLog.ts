import { makeAutoObservable } from "mobx";

export type LogLevel = "info" | "error";

export interface CommandLog {
  level: LogLevel;
  message: string;
}

export class CommandLogger {
  private _logs: CommandLog[] = [];
  private _logCount = 0;

  constructor(public _maxLogs: number = 100) {
    makeAutoObservable(this);
  }

  get logs(): CommandLog[] {
    return this._logs;
  }

  get logCount(): number {
    return this._logCount;
  }

  add(level: LogLevel, message: string) {
    this._logs.push({ level, message });
    this._logs = this._logs.slice(-this._maxLogs);
    this._logCount++;
  }

  info(message: string) {
    this.add("info", message);
  }

  error(message: string) {
    this.add("error", message);
  }

  clear() {
    this._logs = [];
  }

  get maxLogs(): number {
    return this._maxLogs;
  }

  set maxLogs(maxLogs: number) {
    this._maxLogs = maxLogs;
    this._logs = this._logs.slice(-maxLogs);
  }
}
