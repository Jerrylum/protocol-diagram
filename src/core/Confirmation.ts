import { makeAutoObservable, when } from "mobx";
import { Modals } from "./Modals";

export interface ConfirmationButton {
  label: string;
  onClick?: () => void;
  hotkey?: string;
  color?: "inherit" | "primary" | "secondary" | "success" | "error" | "info" | "warning";
}

export interface ConfirmationPromptData {
  title: string;
  description: React.ReactNode;
  buttons: ConfirmationButton[];
  onKeyDown?(e: React.KeyboardEvent<HTMLDivElement>, onClick: (index: number) => void): void;
}

export interface ConfirmationInputPromptData extends ConfirmationPromptData {
  inputLabel: string;
  inputDefaultValue: string;
}

export class Confirmation {
  private data?: ConfirmationPromptData | ConfirmationInputPromptData;
  public input?: string;

  constructor(private modals: () => Modals) {
    makeAutoObservable(this);
  }

  close() {
    this.modals().close(ConfirmationModalSymbol);
    this.data = undefined;
  }

  async prompt(data: ConfirmationPromptData | ConfirmationInputPromptData, priority: number = 1) {
    if (data.buttons.length === 0) {
      data.buttons.push({ label: "OK" });
    }

    this.close();
    this.data = data;
    if ("inputLabel" in data && "inputDefaultValue" in data) {
      this.input = data.inputDefaultValue;
    } else {
      this.input = undefined;
    }

    this.modals().open(ConfirmationModalSymbol, priority);

    await when(() => this.data === undefined);

    return this.input;
  }

  get isOpen() {
    return this.data !== undefined && this.modals().opening === ConfirmationModalSymbol;
  }

  get title() {
    return this.data?.title ?? "";
  }

  set title(value: string) {
    if (this.data === undefined) return;
    this.data.title = value;
  }

  get description() {
    return this.data?.description ?? "";
  }

  set description(value: React.ReactNode) {
    if (this.data === undefined) return;
    this.data.description = value;
  }

  get buttons() {
    return this.data?.buttons ?? [];
  }

  get onKeyDown() {
    return this.data?.onKeyDown;
  }

  get inputLabel() {
    return (this.data as ConfirmationInputPromptData)?.inputLabel ?? "";
  }
}

export const ConfirmationModalSymbol = Symbol("ConfirmationModalSymbol");
