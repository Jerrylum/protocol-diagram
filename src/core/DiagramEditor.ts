import { makeAutoObservable } from "mobx";
import { clamp } from "./Util";
import { Vector } from "./Vector";

export class DiagramEditor {
  private _offset: Vector = new Vector(0, 0);
  private _scale: number = 1; // 1 = 100%, [1..3]

  get offset() {
    return this._offset;
  }

  get scale() {
    return this._scale;
  }

  set offset(offset: Vector) {
    this._offset = offset;
  }

  set scale(scale: number) {
    this._scale = clamp(scale, 0.75, 2);
  }

  constructor() {
    makeAutoObservable(this);
  }

  resetOffsetAndScale() {
    this.offset = new Vector(0, 0);
    this.scale = 1;
  }
}