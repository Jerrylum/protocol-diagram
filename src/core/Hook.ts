import React from "react";

export function useBetterMemo<T extends { destructor: () => void } | {}>(
  factory: () => T,
  deps: React.DependencyList = []
): T {
  const storage = React.useMemo(() => factory(), deps); // eslint-disable-line react-hooks/exhaustive-deps
  React.useEffect(() => {
    return () => {
      if ("destructor" in storage && typeof storage.destructor === "function") storage.destructor?.();
    };
  }, deps); // eslint-disable-line react-hooks/exhaustive-deps

  return storage;
}
