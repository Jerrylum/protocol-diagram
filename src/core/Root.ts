import { Modals } from "./Modals";

export type RootStore = typeof rootStore;

const rootStore = {
  modals: new Modals()
} as const;

export function getRootStore(): RootStore {
  return rootStore;
}

