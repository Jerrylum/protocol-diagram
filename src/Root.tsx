import "./Root.scss";
import { observer } from "mobx-react-lite";
import { DiagramCanvas } from "./app/DiagramCanvas";
import { Box } from "@mui/material";
import { BottomPanel } from "./app/BottomPanel";
// import { TestModal, TestModalSymbol } from "./app/TestModal";
// import React from "react";
// import { getRootStore } from "./core/Root";

const Root = observer(() => {
  // React.useEffect(() => {
  //   getRootStore().modals.open(TestModalSymbol);
  // }, []);
  return (
    <Box id="root-container">
      <DiagramCanvas />
      <BottomPanel />
      {/* <TestModal /> */}
    </Box>
  );
});

export default Root;

