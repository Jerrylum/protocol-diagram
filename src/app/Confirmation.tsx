import { Box, Button, Card, Typography } from "@mui/material";
import React from "react";
import { action, observable, reaction } from "mobx";
import { observer } from "mobx-react-lite";
import { useBetterMemo } from "../core/Hook";
import { ObserverInput } from "../component/ObserverInput";
import { Modal } from "../component/Modal";
import { getRootStore } from "../core/Root";
import { ConfirmationModalSymbol } from "../core/Confirmation";

export const ConfirmationModal = observer(() => {
  const { confirmation: cfm } = getRootStore();

  const buttons = useBetterMemo(() => observable([] as (HTMLButtonElement | null)[]), [cfm.isOpen]);
  const [renderCount, setRenderCount] = React.useState(0);

  React.useEffect(() => {
    return reaction(
      () => buttons.length,
      () => setRenderCount(renderCount => renderCount + 1)
    );
  }, [buttons]);

  React.useEffect(() => {
    (document.querySelector("*[data-testid=sentinelStart]") as HTMLElement | undefined)?.focus();
  }, [renderCount]);

  if (cfm.isOpen === false) return null;

  const focusOnButton = (offset: number) => buttons[buttons.indexOf(document.activeElement as any) + offset]?.focus();

  function onClick(idx: number) {
    if (idx < 0 || idx >= cfm.buttons.length) idx = cfm.buttons.length - 1;

    const func = cfm.buttons[idx]?.onClick;
    cfm.close();
    // ALGO: Call after closing the dialog in case the callback opens another dialog
    func?.();
  }

  function onKeyDown(e: React.KeyboardEvent<HTMLDivElement>) {
    if (e.key === "ArrowLeft") {
      focusOnButton(-1);
    } else if (e.key === "ArrowRight") {
      focusOnButton(1);
    } else {
      const index = cfm.buttons.findIndex(btn => btn.hotkey === e.key);
      index !== -1 && onClick(index);
    }

    cfm.onKeyDown?.(e, onClick);
  }

  function onInputKeyDown(e: React.KeyboardEvent<HTMLInputElement>) {
    if (e.code === "Enter" || e.code === "NumpadEnter") {
      onClick(0);
    }
  }

  return (
    <Modal symbol={ConfirmationModalSymbol} onClose={action(onClick.bind(null, -1))}>
      <Card id="confirmation-modal" className="modal-container" onKeyDown={action(onKeyDown)}>
        <Typography variant="h2" gutterBottom>
          {cfm.title}
        </Typography>
        {/* https://stackoverflow.com/questions/9769587/set-div-to-have-its-siblings-width */}
        <Box sx={{ display: "flex" }}>
          <Typography component="div" variant="body1" gutterBottom sx={{ flexGrow: "1", width: "0" }}>
            {cfm.description}
          </Typography>
        </Box>
        {cfm.input !== undefined && (
          <Box sx={{ width: "100%" }}>
            <ObserverInput
              label={cfm.inputLabel}
              getValue={() => cfm.input!}
              setValue={value => (cfm.input = value)}
              isValidIntermediate={() => true}
              isValidValue={() => true}
              onKeyDown={onInputKeyDown}
              sx={{ width: "100%" }}
            />
          </Box>
        )}
        <Box className="button-box">
          {cfm.buttons.map((btn, i) => {
            return (
              <Button
                key={i}
                disableRipple
                variant="text"
                color={btn.color ?? "inherit"}
                ref={action((element: HTMLButtonElement | null) => (buttons[i] = element))}
                onClick={action(onClick.bind(null, i))}>
                {btn.label}
                {btn.hotkey ? `(${btn.hotkey.toUpperCase()})` : ""}
              </Button>
            );
          })}
        </Box>
      </Card>
    </Modal>
  );
});
