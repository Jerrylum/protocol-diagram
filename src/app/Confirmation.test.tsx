import { act } from "react-dom/test-utils";
import { ConfirmationModal } from "./Confirmation";
import { fireEvent, render } from "@testing-library/react";
import { getRootStore } from "../core/Root";

test("Render Confirmation", () => {
  const { confirmation } = getRootStore();
  confirmation.prompt({
    title: "title1",
    description: "description1",
    buttons: [{ label: "OK", hotkey: "Escape" }],
    onKeyDown: undefined,
    inputLabel: "testLabel",
    inputDefaultValue: "testDefaultValue"
  });
  const components = (
    <div id="root-container">
      <ConfirmationModal />
    </div>
  );
  const result = render(components);
  expect(result.container).toMatchSnapshot();
  let container = document.querySelector(".modal-container");
  act(() => {
    fireEvent.keyDown(container!, { key: "ArrowLeft", code: "ArrowLeft" });
  });
  result.rerender(components);

  act(() => {
    fireEvent.keyDown(container!, { key: "ArrowRight", code: "ArrowRight" });
  });
  result.rerender(components);

  let observerInput = result.container.querySelector("input");
  act(() => {
    fireEvent.input(observerInput!, { target: { value: "test" } });
    fireEvent.keyDown(observerInput!, { key: "Enter", code: "Enter" });
  });
  result.rerender(components);

  confirmation.prompt({
    title: "title1",
    description: "description1",
    buttons: [{ label: "OK", hotkey: "Escape" }],
    onKeyDown: undefined,
    inputLabel: "testLabel",
    inputDefaultValue: "testDefaultValue"
  });

  result.rerender(components);

  container = document.querySelector(".modal-container");
  observerInput = result.container.querySelector("input");

  act(() => {
    fireEvent.input(observerInput!, { target: { value: "test" } });
    fireEvent.keyDown(observerInput!, { key: "NumpadEnter", code: "NumpadEnter" });
  });
  result.rerender(components);

  confirmation.prompt({
    title: "title1",
    description: "description1",
    buttons: [{ label: "OK", hotkey: "Escape" }],
    onKeyDown: undefined,
    inputLabel: "testLabel",
    inputDefaultValue: "testDefaultValue"
  });

  result.rerender(components);

  container = document.querySelector(".modal-container");
  observerInput = result.container.querySelector("input");

  act(() => {
    fireEvent.keyDown(container!, { key: "Enter", code: "Enter" });
  });
  result.rerender(components);

  act(() => {
    fireEvent.keyDown(container!, { key: "Escape", code: "Escape" });
  });
  result.rerender(components);
});
