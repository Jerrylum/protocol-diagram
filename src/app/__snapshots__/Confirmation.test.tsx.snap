// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Confirmation 1`] = `
<div
  aria-hidden="true"
>
  <div
    id="root-container"
  >
    <div
      class="MuiModal-root css-79ws1d-MuiModal-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root modal-backdrop MuiModal-backdrop modal-backdrop css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root modal-container css-bhp9pd-MuiPaper-root-MuiCard-root"
        id="confirmation-modal"
        tabindex="-1"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 MuiTypography-gutterBottom css-i1hrn3-MuiTypography-root"
        >
          title1
        </h2>
        <div
          class="MuiBox-root css-k008qs"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-1p3krao-MuiTypography-root"
          >
            description1
          </div>
        </div>
        <div
          class="MuiBox-root css-8atqhb"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r1:"
              id=":r1:-label"
            >
              testLabel
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                type="text"
                value="testDefaultValue"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    testLabel
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="button-box MuiBox-root css-0"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-1y942vo-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            OK
            (ESCAPE)
          </button>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;
